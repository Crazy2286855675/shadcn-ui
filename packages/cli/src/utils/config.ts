import os from "os"
import path from "path"
import fs from "fs-extra"
import { z } from "zod"

// $HOME/.config/shadcn-ui/config.json
const CONFIG_PATH = path.join(
  os.homedir(),
  ".config",
  "shadcn-ui",
  "config.json"
)

const configSchema = z.object({
  projects: z.record(z.string(), z.string()),
})

type Config = z.infer<typeof configSchema>

// initialize the config
// create the config file if it doesn't exist
/**
 * ---
 * Initialize the config. Ensure that the config file exists and has the corret content
 */
export const init = async () => {
  await fs.ensureFile(CONFIG_PATH)

  const content = await fs.readFile(CONFIG_PATH, "utf-8")

  if (!!content) return
  await fs.writeJson(CONFIG_PATH, { projects: {} }, { spaces: 2 })
}

/**
 * ---
 * Get the config object
 */
export const get = async (): Promise<Config> => {
  const json = await fs.readJson(CONFIG_PATH)
  return configSchema.parse(json)
}

/**
 * ---
 * Upsert a project preference
 *
 * @param project - The current working dir
 * @param value - The path where to store the components
 */
export const upsertProject = async (project: string, value: string) => {
  const config = await get()

  config.projects[project] = value

  await fs.writeJSON(CONFIG_PATH, config, { spaces: 2 })
}
