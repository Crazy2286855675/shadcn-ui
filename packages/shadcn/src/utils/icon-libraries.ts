export const ICON_LIBRARIES = {
  lucide: "lucide-react",
  radix: "@radix-ui/react-icons",
}
