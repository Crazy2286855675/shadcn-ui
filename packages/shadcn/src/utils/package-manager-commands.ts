import { execa } from "execa";
import { getPackageManager, PackageManager } from "./get-package-manager";

export interface Dependencies {
  dependencies?: string[],
  devDependencies?: string[],
}

export interface PackageManagerOptions {
  cwd: string;
}

function dependencyCommand(packageManager: PackageManager, dependencies: string[]) {
  const commands: string[] = [];
  switch (packageManager) {
    case "npm":
      commands.push("install");
      break;
    case "yarn":
    case "bun":
    case "pnpm":
    case "rush":
      commands.push("add");
      break;
  }

  if (packageManager === "rush") {
    // build rush command to install multiple dependencies : reference : https://rushjs.io/pages/commands/rush_add
    dependencies.forEach(packageName=>commands.push('-p',`${packageName}`));
  } else {
    commands.push(...dependencies);
  }
  return commands;
}

function devDependencyCommand(packageManager: PackageManager, devDependencies: string[]) {
  const commands = dependencyCommand(packageManager, devDependencies);

  switch (packageManager) {
    case "npm":
    case "pnpm":
      commands.push('-D');
    break;
    case "yarn":
    case "bun":
    case "rush":
      commands.push('--dev');
      break;
  }
  return commands;
}

export async function installDependencies(dependencies: Dependencies, options:PackageManagerOptions) {

  if ((!dependencies.dependencies || dependencies.dependencies.length === 0) &&
    (!dependencies.devDependencies || dependencies.devDependencies.length === 0)) {
    return;
  }

  const packageManager = await getPackageManager(options.cwd);

  if (dependencies.dependencies && dependencies.dependencies.length > 0) {
    const commands = dependencyCommand(packageManager, dependencies.dependencies);
    await execa(packageManager, commands, { cwd:options.cwd });
  }

  if (dependencies.devDependencies && dependencies.devDependencies.length > 0) {
    const commands = devDependencyCommand(packageManager, dependencies.devDependencies);
    await execa(packageManager, commands, { cwd:options.cwd });
  }

}
