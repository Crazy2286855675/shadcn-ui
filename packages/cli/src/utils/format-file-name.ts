import { CASE_CONVENTION } from "@/src/utils/registry"
import toCamelCase from "lodash.camelcase"
import toKebabCase from "lodash.kebabcase"
import toSnakeCase from "lodash.snakecase"


export function formatFileName(
  fullFileName: string,
  caseConvention: CASE_CONVENTION
) {
  let fileNameWithExt = fullFileName;
  let fileName = fileNameWithExt;
  let path = '';
  let ext = '';

  if (fullFileName.includes('/')) {
    const lastSlashIndex = fullFileName.lastIndexOf('/');
    fileNameWithExt = fullFileName.substring(lastSlashIndex + 1);
    path = fullFileName.substring(0, lastSlashIndex + 1);
  }

  if (fileNameWithExt.includes('.')) {
    const lastDotIndex = fileNameWithExt.lastIndexOf('.');
    fileName = fileNameWithExt.substring(0, lastDotIndex);
    ext = fileNameWithExt.substring(lastDotIndex);
  }

  switch (caseConvention) {
    case CASE_CONVENTION.KEBAB:
      fileName = toKebabCase(fileName)
      break
    case CASE_CONVENTION.PASCAL:
      const camelCase = toCamelCase(fileName)
      fileName = camelCase.charAt(0).toUpperCase() + camelCase.slice(1)
      break
    case CASE_CONVENTION.CAMEL:
      fileName = toCamelCase(fileName)
      break
    case CASE_CONVENTION.SNAKE:
      fileName = toSnakeCase(fileName)
      break
  }

  return `${path}${fileName}${ext}`
}
