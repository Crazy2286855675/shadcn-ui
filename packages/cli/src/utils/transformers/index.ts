import { promises as fs } from "fs"
import { tmpdir } from "os"
import path from "path"
import { Config } from "@/src/utils/get-config"
import { transformRsc } from "@/src/utils/transformers/transform-rsc"
import { Project, type SourceFile } from "ts-morph"

import { transformImport } from "./transform-import"

export type Transformer = (opts: {
  sourceFile: SourceFile
  filename: string
  raw: string
  config: Config
}) => Promise<SourceFile>

const transformers: Transformer[] = [transformImport, transformRsc]

const project = new Project()

async function createTempSourceFile(filename: string) {
  const dir = await fs.mkdtemp(path.join(tmpdir(), "shadcn-"))
  return path.join(dir, filename)
}

export async function transform(filename: string, raw: string, config: Config) {
  const tempFile = await createTempSourceFile(filename)
  const sourceFile = project.createSourceFile(tempFile, raw)

  for (const transformer of transformers) {
    transformer({ sourceFile, filename, raw, config })
  }

  return sourceFile.getFullText()
}
