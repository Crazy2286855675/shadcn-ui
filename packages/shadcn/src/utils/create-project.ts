import path from "path"
import { initOptionsSchema } from "@/src/commands/init"
import { getPackageManager } from "@/src/utils/get-package-manager"
import { highlighter } from "@/src/utils/highlighter"
import { logger } from "@/src/utils/logger"
import { spinner } from "@/src/utils/spinner"
import { execa } from "execa"
import fs from "fs-extra"
import prompts from "prompts"
import { z } from "zod"

export async function createProject(
  options: Pick<z.infer<typeof initOptionsSchema>, "cwd" | "force" | "srcDir">
) {
  options = {
    srcDir: false,
    ...options,
  }

  if (!options.force) {
    const { proceed } = await prompts({
      type: "confirm",
      name: "proceed",
      message: `The path ${highlighter.info(
        options.cwd
      )} does not contain a package.json file. Would you like to start a new ${highlighter.info(
        "Next.js"
      )} project?`,
      initial: true,
    })
    if (!proceed) {
      return {
        projectPath: null,
        projectName: null,
      }
    }
  }

  const packageManager = await getPackageManager(options.cwd)
  
  const { name } = await prompts({
    type: "text",
    name: "name",
    message: "What is your project named?",
    initial: "my-app",
    format: (value: string) => value.trim(),
  })

  const projectPath = path.join(options.cwd, name)

  // Check if path is writable
  try {
    await fs.access(options.cwd, fs.constants.W_OK)
  } catch (error) {
    logger.break()
    logger.error(`The path ${highlighter.info(options.cwd)} is not writable.`)
    logger.error(
      `It is likely you do not have write permissions for this folder or the path ${highlighter.info(
        options.cwd
      )} does not exist.`
    )
    logger.break()
    process.exit(1)
  }

  if (fs.existsSync(path.resolve(projectPath, "package.json"))) {
    logger.break()
    logger.error(
      `A project with the name ${highlighter.info(name)} already exists.`
    )
    logger.error(`Please choose a different name and try again.`)
    logger.break()
    process.exit(1)
  }

  const createSpinner = spinner(
    `Creating a new Next.js project. This may take a few minutes.`
  ).start()

  const args = [
    "--tailwind",
    "--eslint",
    "--typescript",
    "--app",
    options.srcDir ? "--src-dir" : "--no-src-dir",
    "--no-import-alias",
    `--use-${packageManager}`,
  ]

  try {
    // Run create-next-app in dry-run mode first to validate the project name
    await execa(
      "npx",
      ["create-next-app@latest", projectPath, "--dry-run", ...args],
      {
        cwd: options.cwd,
      }
    )

    // If dry-run succeeds, proceed with actual creation
    await execa(
      "npx",
      ["create-next-app@latest", projectPath, "--silent", ...args],
      {
        cwd: options.cwd,
      }
    )
    
    createSpinner?.succeed("Created a new Next.js project successfully.")
  } catch (error) {
    createSpinner?.fail("Failed to create Next.js project")
    logger.break()
    
    if (error instanceof Error) {
      // Extract and format the Next.js error message
      const errorMessage = error.message
      const nextJsError = errorMessage.includes('Invalid project name:') 
        ? errorMessage.split('Invalid project name:')[1].trim()
        : 'Project name validation failed. Please ensure your project name follows Next.js naming conventions.'
      
      logger.error(`Next.js validation error: ${nextJsError}`)
      logger.info("Next.js requires project names to be lowercase, with no spaces.")
    } else {
      logger.error(
        "Something went wrong creating a new Next.js project. Please try again with a different project name."
      )
    }
    
    logger.break()
    process.exit(1)
  }

  return {
    projectPath,
    projectName: name,
  }
}
