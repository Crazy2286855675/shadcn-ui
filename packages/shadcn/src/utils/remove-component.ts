import path from "path"
import { type Config } from "@/src/utils/get-config"
import { handleError } from "@/src/utils/handle-error"
import { logger } from "@/src/utils/logger"
import {
  getRegistryItemFileTargetPath,
  registryResolveItemsTree,
} from "@/src/utils/registry"
import { spinner } from "@/src/utils/spinner"
import { updateComponentJson } from "@/src/utils/updaters/update-component-json"
// import { updateCssVars } from "@/src/utils/updaters/update-css-vars"
import { removeDependencies } from "@/src/utils/updaters/update-dependencies"
import {
  removeFiles,
  resolveTargetDir,
} from "@/src/utils/updaters/update-files"
import prompts from "prompts"
import { z } from "zod"

import { getProjectInfo } from "./get-project-info"
import { registryResolvedItemsTreeSchema } from "./registry/schema"

export async function removeComponents(
  components: string[],
  config: Config,
  options: {
    silent?: boolean
    isNewProject?: boolean
    recursive?: boolean
  }
) {
  options = {
    silent: false,
    isNewProject: false,
    ...options,
  }

  const registrySpinner = spinner(`Checking registry.`, {
    silent: options.silent,
  })?.start()

  //TODO:create endpoint to return just required feilds instead of this
  const tree = await registryResolveItemsTree(components, config)
  if (!tree) {
    registrySpinner?.fail()
    return handleError(new Error("Failed to fetch components from registry."))
  }
  registrySpinner?.succeed()
  await confirmRemoval(tree, config)
  // TODO:Write removeTailwindConfig function

  // TODO:Write removeCssVars function

  await removeDependencies(tree.dependencies, config, {
    silent: options.silent,
  })
  await removeFiles(tree.files, config, {
    silent: options.silent,
  })
  await updateComponentJson(tree.files, config, options, "remove")
  if (tree.docs) {
    logger.info(tree.docs)
  }
}

async function confirmRemoval(
  tree: z.infer<typeof registryResolvedItemsTreeSchema>,
  config: Config
) {
  if (tree.files?.length) {
    logger.log("files :")
    const projectInfo = await getProjectInfo(config.resolvedPaths.cwd)

    for (const file of tree.files) {
      const fileName = path.basename(file.path)

      let targetDir = getRegistryItemFileTargetPath(file, config)
      let filePath = path.join(targetDir, fileName)
      if (file.target) {
        filePath = resolveTargetDir(projectInfo, config, file.target)
        targetDir = path.dirname(filePath)
      }
      const relativePath = path.relative(config.resolvedPaths.cwd, filePath)
      logger.log(`  - ${relativePath}`)
    }
  }
  if (tree.dependencies?.length) {
    logger.log("dependecies :")
    for (const dependency of tree.dependencies) {
      logger.log(`  - ${dependency}`)
    }
  }

  if (tree.devDependencies?.length) {
    logger.log("Dev dependecies :")
    for (const dependency of tree.devDependencies) {
      logger.log(`  - ${dependency}`)
    }
  }

  const { proceed } = await prompts({
    type: "confirm",
    name: "proceed",
    message: "are you sure you want to remove above ?",
    initial: true,
  })

  if (!proceed) {
    logger.info("exiting...")
    logger.break()
    process.exit(1)
  }
}
