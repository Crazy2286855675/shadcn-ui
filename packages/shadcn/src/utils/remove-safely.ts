import { promises as fs, rmSync } from "fs"
import path from "path"
import { type Config } from "@/src/utils/get-config"
import { logger } from "@/src/utils/logger"
import { spinner } from "@/src/utils/spinner"

export async function removeComponentsSafely(
  components: string[],
  config: Config,
  silent: boolean
) {
  const filesDeletedSpinner = spinner(`removing files.`, {
    silent,
  })?.start()
  const filesDeleted: string[] = []
  const filesFailed: string[] = []
  let { components: Component, hooks, lib, ui } = config.installed
  // convert each to set
  const componentSet = new Set(Component)
  const hooksSet = new Set(hooks)
  const libSet = new Set(lib)
  const uiSet = new Set(ui)

  const cwd = config.resolvedPaths.cwd
  let target: string = ""
  components.forEach(async (component) => {
    let file = component + (config.tsx ? ".tsx" : "jsx")
    let relativePath = ""
    let filePath = ""
    // check where is file located
    if (componentSet?.has(component)) {
      target = config.resolvedPaths.components
      componentSet.delete(component)
      filePath = path.join(target, file)
      relativePath = path.relative(cwd, filePath)
    } else if (uiSet?.has(component)) {
      target = config.resolvedPaths.ui
      uiSet.delete(component)
      filePath = path.join(target, file)
      relativePath = path.relative(cwd, filePath)
    } else if (libSet?.has(component)) {
      file = component + (config.tsx ? ".ts" : "js")
      target = config.resolvedPaths.lib
      libSet.delete(component)
      filePath = path.join(target, file)
      relativePath = path.relative(cwd, filePath)
    } else if (hooksSet?.has(component)) {
      target = config.resolvedPaths.hooks
      hooksSet.delete(component)
      filePath = path.join(target, file)
      relativePath = path.relative(cwd, filePath)
    } else {
      filePath = path.join(target, file)
      relativePath = path.relative(cwd, filePath)
      filesFailed.push(relativePath)
      return
    }
    filePath = path.join(target, file)
    relativePath = path.relative(cwd, filePath)
    const success = removeFile(file, target)
    if (success) {
      filesDeleted.push(relativePath)
      return
    }
    filesFailed.push(relativePath)
    return
  })

  filesDeletedSpinner.succeed()
  if (filesFailed.length) {
    spinner(
      `failed to delete ${filesFailed.length} ${
        filesFailed.length === 1 ? "file" : "files"
      }:`
    )?.fail()
    if (!silent) {
      for (const file of filesFailed) {
        logger.log(`  - ${file}`)
      }
    }
  }
  spinner(
    `deleted ${filesDeleted.length} ${
      filesDeleted.length === 1 ? "file" : "files"
    }:`
  )?.succeed()
  if (!silent) {
    for (const file of filesDeleted) {
      logger.log(`  - ${file}`)
    }
  }

  // convert back to array

  const updateSpinner = spinner(`Updating components.json.`, {
    silent: silent,
  })?.start()
  const { resolvedPaths, ...cleanedConfig } = config

  cleanedConfig.installed = {
    components: Array.from(componentSet),
    ui: Array.from(uiSet),
    lib: Array.from(libSet),
    hooks: Array.from(hooksSet),
  }
  // Write the updated config to components.json

  const targetPath = path.resolve(config.resolvedPaths.cwd, "components.json")
  await fs.writeFile(targetPath, JSON.stringify(cleanedConfig, null, 2), "utf8")
  updateSpinner.succeed()
}

function removeFile(fileName: string, resolvedPath: string): boolean {
  try {
    let filePath = path.join(resolvedPath, fileName)
    rmSync(filePath)
    return true
  } catch (error: any) {
    logger.info(error.message)
    return false
  }
}
