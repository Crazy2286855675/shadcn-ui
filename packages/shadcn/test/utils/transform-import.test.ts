import { expect, test } from "vitest"

import { transform } from "../../src/utils/transformers"

test("transform import", async () => {
  expect(
    await transform({
      filename: "test.ts",
      raw: `import * as React from "react"
import { Foo } from "bar"
    import { Button } from "@/registry/new-york/ui/button"
    import { Label} from "ui/label"
    import { Box } from "@/registry/new-york/box"

    import { cn } from "@/lib/utils"
    `,
      config: {
        tsx: true,
        tailwind: {
          baseColor: "neutral",
          cssVariables: true,
        },
        aliases: {
          components: "@/components",
          utils: "@/lib/utils",
        },
      },
    })
  ).toMatchSnapshot()

  expect(
    await transform({
      filename: "test.ts",
      raw: `import * as React from "react"
import { Foo } from "bar"
    import { Button } from "@/registry/new-york/ui/button"
    import { Label} from "ui/label"
    import { Box } from "@/registry/new-york/box"

    import { cn, foo, bar } from "@/lib/utils"
    import { bar } from "@/lib/utils/bar"
    `,
      config: {
        tsx: true,
        aliases: {
          components: "~/src/components",
          utils: "~/lib/utils",
        },
      },
    })
  ).toMatchSnapshot()

  expect(
    await transform({
      filename: "test.ts",
      raw: `import * as React from "react"
import { Foo } from "bar"
    import { Button } from "@/registry/new-york/ui/button"
    import { Label} from "ui/label"
    import { Box } from "@/registry/new-york/box"

    import { cn } from "@/lib/utils"
    import { bar } from "@/lib/utils/bar"
    `,
      config: {
        tsx: true,
        aliases: {
          components: "~/src/components",
          utils: "~/src/utils",
        },
      },
    })
  ).toMatchSnapshot()

  expect(
    await transform({
      filename: "test.ts",
      raw: `import * as React from "react"
import { Foo } from "bar"
    import { Button } from "@/registry/new-york/ui/button"
    import { Label} from "ui/label"
    import { Box } from "@/registry/new-york/box"

    import { cn } from "@/lib/utils"
    import { bar } from "@/lib/utils/bar"
    `,
      config: {
        tsx: true,
        aliases: {
          components: "~/src/components",
          utils: "~/src/utils",
          ui: "~/src/components",
        },
      },
    })
  ).toMatchSnapshot()

  expect(
    await transform({
      filename: "test.ts",
      raw: `import * as React from "react"
import { Foo } from "bar"
    import { Button } from "@/registry/new-york/ui/button"
    import { Label} from "ui/label"
    import { Box } from "@/registry/new-york/box"

    import { cn } from "@/lib/utils"
    import { bar } from "@/lib/utils/bar"
    `,
      config: {
        tsx: true,
        aliases: {
          components: "~/src/components",
          utils: "~/src/lib/utils",
          ui: "~/src/ui",
        },
      },
    })
  ).toMatchSnapshot()

  expect(
    await transform({
      filename: "test.ts",
      raw: `import * as React from "react"
import { Foo } from "bar"
    import { Button } from "@/registry/new-york/ui/button"
    import { Label} from "ui/label"
    import { Box } from "@/registry/new-york/box"

    import { cn } from "@/lib/utils"
    `,
      config: {
        tsx: true,
        aliases: {
          components: "#app/components",
          utils: "#app/utils",
          ui: "#app/ui",
        },
      },
    })
  ).toMatchSnapshot()

  expect(
    await transform({
      filename: "test.ts",
      raw: `import * as React from "react"
import { Foo } from "bar"
    import { Button } from "@/registry/new-york/ui/button"
    import { Label} from "ui/label"
    import { Box } from "@/registry/new-york/box"

    import { cn } from "@/lib/utils"
    `,
      config: {
        tsx: true,
        aliases: {
          components: "#app/components",
          utils: "~/somepahte/file",
          ui: "#app/ui",
        },
      },
    })
  ).toMatchSnapshot()

  expect(
    await transform({
      filename: "test.ts",
      raw: `import Link from "next/link"

import { Button } from "@/registry/default/ui/button"
import {
  Card,
  CardContent,
  CardDescription,
  CardHeader,
  CardTitle,
} from "@/registry/default/ui/card"
import { Input } from "@/registry/default/ui/input"
import { Label } from "@/registry/default/ui/label"
    `,
      config: {
        tsx: true,
        aliases: {
          components: "#app/components",
          utils: "#app/utils",
        },
      },
    })
  ).toMatchSnapshot()

  expect(
    await transform({
      filename: "test.ts",
      raw: `import { LoginForm } from "@/registry/default/block/login-01/components/login-form"`,
      config: {
        tsx: true,
        aliases: {
          components: "#app/components",
        },
      },
    })
  ).toMatchSnapshot()
})
