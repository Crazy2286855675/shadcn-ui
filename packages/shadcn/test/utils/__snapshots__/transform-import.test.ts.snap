// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`transform import 1`] = `
"import * as React from "react"
import { Foo } from "bar"
    import { Button } from "@/components/ui/button"
    import { Label} from "ui/label"
    import { Box } from "@/components/box"

    import { cn } from "@/lib/utils"
    "
`;

exports[`transform import 2`] = `
"import * as React from "react"
import { Foo } from "bar"
    import { Button } from "~/src/components/ui/button"
    import { Label} from "ui/label"
    import { Box } from "~/src/components/box"

    import { cn, foo, bar } from "~/lib/utils"
    import { bar } from "~/lib/utils/bar"
    "
`;

exports[`transform import 3`] = `
"import * as React from "react"
import { Foo } from "bar"
    import { Button } from "~/src/components/ui/button"
    import { Label} from "ui/label"
    import { Box } from "~/src/components/box"

    import { cn } from "~/src/utils"
    import { bar } from "~/src/utils/bar"
    "
`;

exports[`transform import 4`] = `
"import * as React from "react"
import { Foo } from "bar"
    import { Button } from "~/src/components/button"
    import { Label} from "ui/label"
    import { Box } from "~/src/components/box"

    import { cn } from "~/src/utils"
    import { bar } from "~/src/utils/bar"
    "
`;

exports[`transform import 5`] = `
"import * as React from "react"
import { Foo } from "bar"
    import { Button } from "~/src/ui/button"
    import { Label} from "ui/label"
    import { Box } from "~/src/components/box"

    import { cn } from "~/src/lib/utils"
    import { bar } from "~/src/lib/utils/bar"
    "
`;

exports[`transform import 6`] = `
"import * as React from "react"
import { Foo } from "bar"
    import { Button } from "#app/ui/button"
    import { Label} from "ui/label"
    import { Box } from "#app/components/box"

    import { cn } from "#app/utils"
    "
`;

exports[`transform import 7`] = `
"import * as React from "react"
import { Foo } from "bar"
    import { Button } from "#app/ui/button"
    import { Label} from "ui/label"
    import { Box } from "#app/components/box"

    import { cn } from "~/somepahte/file"
    "
`;

exports[`transform import 8`] = `
"import Link from "next/link"

import { Button } from "#app/components/ui/button"
import {
  Card,
  CardContent,
  CardDescription,
  CardHeader,
  CardTitle,
} from "#app/components/ui/card"
import { Input } from "#app/components/ui/input"
import { Label } from "#app/components/ui/label"
    "
`;

exports[`transform import 9`] = `"import { LoginForm } from "#app/components/login-form""`;
